sql/sys_vars.cc

```diff
@@ -3730,6 +3731,44 @@ static Sys_var_charptr Sys_time_format(
        NO_MUTEX_GUARD, NOT_IN_BINLOG, ON_CHECK(0), ON_UPDATE(0),
        DEPRECATED(""));
 
+/**
+  Pre-update function to commit connection's active transactions when autocommit
+  is enabled.
+
+  @note This hook relies on the fact that it is called while not holding any
+        locks. Breaking this assumption might result in deadlocks as commit
+        acquires many different locks in its process (e.g. to open GTID-related
+        tables).
+
+  @param[in] self   A pointer to the sys_var, i.e. Sys_autocommit.
+  @param[in] thd    A reference to THD object.
+  @param[in] var    A pointer to the set_var created by the parser.
+
+  @retval true   Error during commit
+  @retval false  Otherwise
+*/
+static bool pre_autocommit(sys_var *self, THD *thd, set_var *var)
+{
+  if (!(var->type == OPT_GLOBAL) &&
+      (thd->variables.option_bits & OPTION_NOT_AUTOCOMMIT) &&
+       var->save_result.ulonglong_value)
+  {
+    // Autocommit mode is about to be activated.
+#ifdef WITH_WSREP
+    if (trans_commit_stmt(thd) || trans_commit(thd))
+    {
+      thd->mdl_context.release_transactional_locks();
+      WSREP_DEBUG("autocommit, MDL TRX lock released: %lu", thd->thread_id);
+      return true;
+    }
+#else
+    if (trans_commit_stmt(thd) || trans_commit(thd))
+      return true;
+#endif
```

Comment:
Here we have a common `if (trans_commit_stmt(thd) || trans_commit(thd))` condition in both `#ifdef WITH_WSREP` and `#else` part, which I think can be put outside the `#ifdef` blocks. Can you try to make both look alike?

Note: In 5.7, we have this as
https://github.com/percona/percona-xtradb-cluster/blob/Percona-XtraDB-Cluster-5.7.35-31.53/sql/sys_vars.cc#L5054-L5065

Reply:
That was my 1st thought as well, but because the above mentioned `if` does not use curly brackets in original version, the result will be hard to read like:
```
    if (trans_commit_stmt(thd) || trans_commit(thd))
#ifdef WITH_WSREP
    {
      thd->mdl_context.release_transactional_locks();
      WSREP_DEBUG("autocommit, MDL TRX lock released: %lu", thd->thread_id);
#endif
      return true;
#ifdef WITH_WSREP
    }
#endif
```

Reply:
How about this 
```C++
    if (trans_commit_stmt(thd) || trans_commit(thd))
#ifdef WITH_WSREP
    {
      thd->mdl_context.release_transactional_locks();
      WSREP_DEBUG("Transaction commit failed while toggling autocommit."
                  " Release MDL trx lock for thread: %u", thd->thread_id());
      return true;
    }
#else
      return true;
#endif /* WITH_WSREP */
```

Reply:
Agree.